File: Src/Xt.Synth0.DSP/PatternDSP.cs
using Xt.Synth0.Model;

namespace Xt.Synth0.DSP
{
	class PatternDSP
	{
		internal void Automate(SynthModel synth,
			SequencerModel seq, AudioModel audio, bool[] automated)
		{
			var row = seq.Pattern.Rows[audio.CurrentRow];
			foreach (var fx in row.Fx)
				Automate(synth, fx, automated);
		}

		void Automate(SynthModel model, PatternFx fx, bool[] automated)
		{
			var autos = model.AutoParams();
			var target = fx.Target.Value;
			if (target == 0 || target > autos.Count) return;
			var param = autos[target - 1].Param;
			var value = fx.Value.Value;
			if (value < param.Info.Min) param.Value = param.Info.Min;
			else if (value > param.Info.Max) param.Value = param.Info.Max;
			else param.Value = value;
			automated[target - 1] = true;
		}
	}
}

File: Src/Xt.Synth0.DSP.Native/DSP/PatternDSP.cpp
#include "PatternDSP.hpp"

namespace Xts {

void
PatternDSP::Automate(EditModel const& edit, PatternRow const& row, SynthModel& synth) const
{
  for(int f = 0; f < edit.fx; f++)
    Automate(row.fx[f], synth);
}

void 
PatternDSP::Automate(PatternFx const& fx, SynthModel& synth) const
{
		if (fx.target == 0 || fx.target > TrackConstants::ParamCount) return;
    auto const& p = synth.params[fx.target - 1];
		if (fx.value < p.min) *p.value = p.min;
		else if (fx.value > p.max) *p.value = p.max;
		else *p.value = fx.value;
}

} // namespace Xts